sklearn/datasets/tests/test_openml.py

```diff
@@ -0,0 +1,120 @@
+"""Test the openml loader.
+"""
+import json
+import numpy as np
+import sklearn
+
+from sklearn.datasets import fetch_openml
+from sklearn.datasets.openml import _get_data_features
+from sklearn.utils.testing import (assert_warns_message,
+                                   assert_raise_message)
+
+
+def fetch_dataset_from_openml(data_id, data_name, data_version, expected_observations, expected_features):
+    # fetch with version
+    data_by_name_id = fetch_openml(data_name, version=data_version)
+    assert int(data_by_name_id.details['id']) == data_id
+
+    # fetch without version
+    fetch_openml(data_name)
+    # without specifying the version, there is no guarantee that the data id will be the same
+
+    # fetch with dataset id
+    data_by_id = fetch_openml(data_id)
+    assert data_by_id.details['name'] == data_name
+    assert data_by_id.data.shape == (expected_observations, expected_features)
+    assert data_by_id.target.shape == (expected_observations, )
+
+    # check numeric features:
+    feature_name_type = {feature['name']: feature['data_type'] for feature in _get_data_features(data_id)}
+    for idx, feature_name in enumerate(data_by_id.feature_names):
+        if feature_name_type[feature_name] == 'numeric':
+            # casting trick according to Jaime at
+            # stackoverflow.com/questions/19486283/how-do-i-quickly-check-if-all-elements-of-numpy-array-are-floats
+            assert np.issubdtype(np.array(list(data_by_id.data[:, idx])).dtype, np.number)
+
+    if 'default_target_attribute' in data_by_id.details:
+        target = data_by_id.details['default_target_attribute']
+        if feature_name_type[target] == 'numeric':
+            assert np.issubdtype(np.array(list(data_by_id.data[:, idx])).dtype, np.number)
+    return data_by_id
+
+
+def mock_data_description(id):
+    description = open('mock_openml/%d/data_description.txt' % id, 'r').read()
+    return json.loads(description)['data_set_description']
+
+
+def mock_data_features(id):
+    features = open('mock_openml/%d/data_features.txt' % id, 'r').read()
+    return json.loads(features)['data_features']['feature']
+
+
+def _patch_webbased_functions():
+    sklearn.datasets.openml._get_data_description_by_id = mock_data_description
+    sklearn.datasets.openml._get_data_features = mock_data_features
```

Comment:
It would be better to use pytest monkeypatching https://docs.pytest.org/en/latest/monkeypatch.html